pkg/devfile/adapters/common/command.go

```diff
@@ -13,66 +13,70 @@ import (
 
-	for _, command := range commands {
-
-		// validate command
-		err = validateCommand(data, command)
-
+	if commandName == "" {
+		// validate the command groups before searching for a command match
+		// if the command groups are invalid, err out
+		// we only validate when the push command flags are absent,
+		// since we know the command kind from the push flags
+		err = validateCommandsForGroup(data, groupType)
 		if err != nil {
 			return common.DevfileCommand{}, err
 		}
+	}
 
+	for _, command := range commands {
 		// if command is specified via flags, it has the highest priority
 		// search through all commands to find the specified command name
 		// if not found fallback to error.
 		if commandName != "" {
 
-			// Update Group only custom commands (specified by odo flags)
-			command = updateGroupforCommand(groupType, command)
-
 			if command.Exec.Id == commandName {
 
+				// Update Group only custom commands (specified by odo flags)
+				command = updateCommandGroupIfReqd(groupType, command)
+
 				// we have found the command with name, its groupType Should match to the flag
 				// e.g --build-command "mybuild"
 				// exec:
 				//   id: mybuild
 				//   group:
 				//     kind: build
 				if command.Exec.Group.Kind != groupType {
-					return supportedCommand, fmt.Errorf("command group mismatched, command %s is of group %v in devfile.yaml", commandName, command.Exec.Group.Kind)
+					return command, fmt.Errorf("command group mismatched, command %s is of group %v in devfile.yaml", commandName, command.Exec.Group.Kind)
 				}
-				supportedCommand = command
-				return supportedCommand, nil
+
+				return command, validateCommand(data, command)
 			}
 			continue
 		}
 
 		// if no command specified via flag, default command has the highest priority
-		// We need to scan all the commands to find default command
 		// exec.Group is a pointer, to avoid null pointer
-		if command.Exec.Group != nil && command.Exec.Group.Kind == groupType && command.Exec.Group.IsDefault {
-			supportedCommand = command
-			return supportedCommand, nil
+		if command.Exec.Group != nil && command.Exec.Group.Kind == groupType {
+			if command.Exec.Group.IsDefault {
+				// We need to scan all the commands to find default command
+				return command, validateCommand(data, command)
+			} else if reflect.DeepEqual(firstMatchedCommand, common.DevfileCommand{}) {
+				// store the first matched command in case there is no default command
+				firstMatchedCommand = command
+			}
 		}
 	}
 
 	if commandName == "" {
 		// if default command is not found return the first command found for the matching type.
-		for _, command := range commands {
-
-			if command.Exec.Group != nil && command.Exec.Group.Kind == groupType {
-				supportedCommand = command
-				return supportedCommand, nil
-			}
+		if !reflect.DeepEqual(firstMatchedCommand, common.DevfileCommand{}) {
+			return firstMatchedCommand, validateCommand(data, firstMatchedCommand)
```

Comment:
IMO this is not required now, this block could be removed and we can return from above `else if` block

Reply:
Actually we need this. Prior to this PR, we used two for loops to get the default command and the first command of group type.

But now we're checking for both in the same for loop..

For example if a devfile has two build commands in order
- devbuild1 
- devbuild2 {default:true}

if we removed the if condition and returned in `else if`, it will incorrectly return devbuild1 even tho in the next iteration there is a default command.

But with this if block it solves the two cases ie multiple commands of same group in any order and also one command of a group kind.